### PR TITLE
Jgfouca/upstream merge 2022 12 09

### PR DESCRIPTION
**Pull request by jgfouca**
Conflicts:
components/eam/src/physics/rrtmgp/external (submodule)

Solution: take upstream?

### CODE REVIEW

**Comment by AaronDonahue** [comment]
I think @brhillman  should make the ultimate judgement since the only only conflict is w/ regards to rrtmgp.  Thanks for putting this in @jgfouca

**Comment by ambrad** [comment]
I'm going to run SCREAMv1 tests on Ascent with this PR, so don't merge until I report back.

**Comment by ambrad** [comment]
I'm getting this build error on Ascent in e3sm.bldlog:
```
-- ** Building YAKL for a CUDA backend **
-- ** YAKL is using the following C++ flags: -DYAKL_ARCH_CUDA --expt-extended-lambda --expt-relaxed-constexpr -Wno-deprecated-gpu-targets -std=c++17  **
CMake Error at eamxx/src/physics/rrtmgp/CMakeLists.txt:31 (target_link_libraries):
  The plain signature for target_link_libraries has already been used with
  the target "yakl".  All uses of target_link_libraries with a target must be
  either all-keyword or all-plain.

  The uses of the plain signature are here:

   * /autofs/nccsopen-svm1_home/ambradl/repo/SCREAMref/externals/YAKL/CMakeLists.txt:43 (target_link_libraries)
```

**Comment by bartgol** [comment]
> I'm getting this build error on Ascent in e3sm.bldlog:
> 
> ```
> -- ** Building YAKL for a CUDA backend **
> -- ** YAKL is using the following C++ flags: -DYAKL_ARCH_CUDA --expt-extended-lambda --expt-relaxed-constexpr -Wno-deprecated-gpu-targets -std=c++17  **
> CMake Error at eamxx/src/physics/rrtmgp/CMakeLists.txt:31 (target_link_libraries):
>   The plain signature for target_link_libraries has already been used with
>   the target "yakl".  All uses of target_link_libraries with a target must be
>   either all-keyword or all-plain.
> 
>   The uses of the plain signature are here:
> 
>    * /autofs/nccsopen-svm1_home/ambradl/repo/SCREAMref/externals/YAKL/CMakeLists.txt:43 (target_link_libraries)
> ```
> 
> Edit: I triple-checked my submodules and they are all up to date.

I got that error while working on #2091. I fixed it in there. You can try to manually merge that branch, and see if things get fixed.

**Comment by AaronDonahue** [comment]
@brhillman when you get a chance can you check this?  @ambrad did you have a chance to try @bartgol suggestion on ascent?

**Comment by AaronDonahue** [comment]
Removing automerge until Andrew reports back that things are working on ascent.

**Comment by jgfouca** [comment]
@ambrad , on it.

**Comment by jgfouca** [comment]
Unfortunately, it looks like the build errors remain. Neither submodule set to upstream or scream/master works. I will dig a bit to see if it's an easy fix.

**Comment by jgfouca** [comment]
@brhillman , see latest commit for my attempt to adapt rrtmgp to latest YAKL.

File: components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp [hunk 1]
```diff
@@ -150,7 +150,7 @@ namespace scream {
          */
         template<class S, class T> void limit_to_bounds(S const &arr_in, T const lower, T const upper, S &arr_out) {
             yakl::c::parallel_for(arr_in.totElems(), YAKL_LAMBDA(int i) {
-                arr_out.data()[i] = min(max(arr_in.data()[i], lower), upper);
+                arr_out.data()[i] = std::min(std::max(arr_in.data()[i], lower), upper);
```
Comment (jgfouca) [inline components/eamxx/src/physics/rrtmgp/scream_rrtmgp_interface.hpp:153]: @brhillman , based on YAKL changes and other changes from master, I think this min and max must have been referring to the ones in `std::`

**Comment by ambrad** [comment]
Looks like this should do the trick. I'll run a case on Ascent, and if that works, push this change.
```diff
diff --git a/components/eamxx/src/physics/rrtmgp/CMakeLists.txt b/components/eamxx/src/physics/rrtmgp/CMakeLists.txt
index b2b5d96629..ced6e996c3 100644
--- a/components/eamxx/src/physics/rrtmgp/CMakeLists.txt
+++ b/components/eamxx/src/physics/rrtmgp/CMakeLists.txt
@@ -35,7 +35,7 @@ else ()

   # EAMxx *requires* MPI, so simply look for it, then link against it
   find_package(MPI REQUIRED COMPONENTS C)
-  target_link_libraries (yakl PUBLIC MPI::MPI_C)
+  target_link_libraries (yakl MPI::MPI_C)
   EkatDisableAllWarning(yakl)

   # For debug builds, set -DYAKL_DEBUG
```

**Comment by jgfouca** [comment]
@ambrad , I think we may be waiting on this PR: https://github.com/E3SM-Project/scream/pull/1704

**Comment by ambrad** [comment]
Ok, cool. I'll drop this, then.

**Comment by ambrad** [comment]
@jgfouca we can close this, now, right, since the RRTMGPXX PR has this in it?

**Comment by jgfouca** [comment]
Yep.

**Pull request closed without merging**